File: leetcode/accounts_merge/accounts_merge.cpp
/*
 * Time Complexity: O(NK log(NK))
 *   - N is the number of accounts, K is the maximum number of emails per account.
 *   - Each email is processed, and sorting dominates the complexity.
 *
 * Space Complexity: O(NK)
 *   - For storing parent, size, email mappings, and merged results.
 */

#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using std::vector;
using std::string;
using std::unordered_map;

struct DisjointSet {
    vector<int> parent, size;
    DisjointSet(int n) : parent(n), size(n, 1) {
        for(int i = 0; i < n; ++i) parent[i] = i;
    }
    int find(int node) {
        if(parent[node] == node) return node;
        return parent[node] = find(parent[node]);
    }
    void unionBySize(int u, int v) {
        int pu = find(u), pv = find(v);
        if(pu == pv) return;
        if(size[pu] < size[pv]) {
            parent[pu] = pv;
            size[pv] += size[pu];
        } else {
            parent[pv] = pu;
            size[pu] += size[pv];
        }
    }
};

class Solution {
public:
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        int n = accounts.size();
        DisjointSet ds(n);
        unordered_map<string, int> emailToIndex;

        // Union accounts by shared emails
        for(int i = 0; i < n; ++i) {
            for(int j = 1; j < accounts[i].size(); ++j) {
                const string& email = accounts[i][j];
                if(emailToIndex.count(email)) {
                    ds.unionBySize(i, emailToIndex[email]);
                } else {
                    emailToIndex[email] = i;
                }
            }
        }

        // Group emails by their root parent
        vector<vector<string>> mergedEmails(n);
        for(const auto& [email, idx] : emailToIndex) {
            int root = ds.find(idx);
            mergedEmails[root].push_back(email);
        }

        // Build the answer
        vector<vector<string>> ans;
        for(int i = 0; i < n; ++i) {
            if(mergedEmails[i].empty()) continue;
            std::sort(mergedEmails[i].begin(), mergedEmails[i].end());
            vector<string> entry = {accounts[i][0]};
            entry.insert(entry.end(), mergedEmails[i].begin(), mergedEmails[i].end());
            ans.push_back(entry);
        }
        return ans;
    }
};

File: leetcode/accounts_merge/accounts_merge.java
package accounts_merge;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class DisjoinSet {
    List<Integer> parent = new ArrayList<>();
    List<Integer> size = new ArrayList<>();

    public DisjoinSet(int n) {
        for(int i = 0; i < n; i++) {
            parent.add(i);
            size.add(1);
        }
    }

    public int findUnionParent(int node) {
        if(node == parent.get(node)) return node;

        int unionParent = findUnionParent(parent.get(node));
        parent.set(node, unionParent);
        return parent.get(node); 
    }

    public void unionBySize(int u, int v) {
        int unionParentU = findUnionParent(u);
        int unionParentV = findUnionParent(v);

        if(unionParentU == unionParentV) return;

        int newSize = size.get(unionParentU) + size.get(unionParentV);
        if(size.get(unionParentV) < size.get(unionParentU)) {
            parent.set(unionParentV, unionParentU);
            size.set(unionParentU, newSize);
        } else {
            parent.set(unionParentU, unionParentV);
            size.set(unionParentV, newSize);
        }
    }

}

class Solution {
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        int n = accounts.size();
        DisjoinSet ds = new DisjoinSet(n);
        HashMap<String, Integer> accountIndexByEmail = new HashMap<>();

        for(int i = 0; i < n; i++) {
            for(int j = 1; j < accounts.get(i).size(); j++) {
                String email = accounts.get(i).get(j);
                if(accountIndexByEmail.containsKey(email)) {
                    ds.unionBySize(i, accountIndexByEmail.get(email));
                } else {
                    accountIndexByEmail.put(email, i);
                }
            }
        }

        ArrayList<String>[] mergeEmail = new ArrayList[n];
        for(int i = 0; i < n; i++) {
            mergeEmail[i] = new ArrayList<String>();
        }

        for(Map.Entry<String, Integer> it : accountIndexByEmail.entrySet()) {
            String email = it.getKey();
            int node = ds.findUnionParent(it.getValue());
            mergeEmail[node].add(email);
        }

        List<List<String>> ans = new ArrayList<>();
        for(int i = 0; i < n; i++) {
            if(mergeEmail[i].size() == 0) continue;

            Collections.sort(mergeEmail[i]);

            List<String> tmp = new ArrayList<>();
            tmp.add(accounts.get(i).get(0));
            for(String email : mergeEmail[i]) {
                tmp.add(email);
            }
            ans.add(tmp);
        }

        return ans;
    }
}
